circup/backends.py

```diff
@@ -0,0 +1,806 @@
+        auth = HTTPBasicAuth("", url.password)
+        with self.session.put(library_path, auth=auth, timeout=self.timeout) as r:
+            r.raise_for_status()
+
+    def _install_module_mpy(self, bundle, metadata):
+        """
+        :param bundle library bundle.
+        :param library_path library path
+        :param metadata dictionary.
+        """
+        module_name = os.path.basename(metadata["path"]).replace(".py", ".mpy")
+        if not module_name:
+            # Must be a directory based module.
+            module_name = os.path.basename(os.path.dirname(metadata["path"]))
+        major_version = self.get_circuitpython_version()[0].split(".")[0]
+        bundle_platform = "{}mpy".format(major_version)
+        bundle_path = os.path.join(bundle.lib_dir(bundle_platform), module_name)
+        if os.path.isdir(bundle_path):
+
+            self.install_dir_http(bundle_path)
+
+        elif os.path.isfile(bundle_path):
+            self.install_file_http(bundle_path)
+
+        else:
+            raise IOError("Cannot find compiled version of module.")
+
+    # pylint: enable=too-many-locals,too-many-branches
+    def _install_module_py(self, metadata):
+        """
+        :param library_path library path
+        :param metadata dictionary.
+        """
+
+        source_path = metadata["path"]  # Path to Python source version.
+        if os.path.isdir(source_path):
+            self.install_dir_http(source_path)
+
+        else:
+            self.install_file_http(source_path)
+
+    def get_auto_file_path(self, auto_file_path):
+        """
+        Make a local temp copy of the --auto file from the device.
+        Returns the path to the local copy.
+        """
+        url = auto_file_path
+        auth = HTTPBasicAuth("", self.password)
+        with self.session.get(url, auth=auth, timeout=self.timeout) as r:
+            r.raise_for_status()
+            with open(LOCAL_CODE_PY_COPY, "w", encoding="utf-8") as f:
+                f.write(r.text)
+        return LOCAL_CODE_PY_COPY
+
+    def uninstall(self, device_path, module_path):
+        """
+        Uninstall given module on device using REST API.
+        """
+        url = urlparse(device_path)
+        auth = HTTPBasicAuth("", url.password)
+        with self.session.delete(module_path, auth=auth, timeout=self.timeout) as r:
+            r.raise_for_status()
+
+    def update(self, module):
+        """
+        Delete the module on the device, then copy the module from the bundle
+        back onto the device.
+
+        The caller is expected to handle any exceptions raised.
+        """
+        self._update_http(module)
+
+    def _update_http(self, module):
+        """
+        Update the module using web workflow.
+        """
+        if module.file:
+            # Copy the file (will overwrite).
+            self.install_file_http(module.bundle_path)
+        else:
+            # Delete the directory (recursive) first.
```

Comment:
It's fine as is, but if the device is not writeable and you call `update` it fails with http status code 409. I found this running `update --all` and updating the toml library. Could instead check if writable with call to self.get_free_space() which now bails nicely informing the user the device is not writable. I forgot about the update path when adding the graceful error message.

Reply:
@FoamyGuy

Reply:
Thank you for testing! 

The latest commit adds checks to output a mroe understandable error when the drive is not writeable.